compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -373,6 +321,149 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
         }
     }
 
+    fn suggest_borrow_fn_like(
+        &self,
+        err: &mut DiagnosticBuilder<'tcx, ErrorGuaranteed>,
+        ty: Ty<'tcx>,
+        move_sites: &[MoveSite],
+        value_name: &str,
+    ) -> bool {
+        let tcx = self.infcx.tcx;
+
+        // Find out if the predicates show that the type is a Fn or FnMut
+        let find_fn_kind_from_did = |predicates: &[(ty::Predicate<'tcx>, Span)], substs| {
+            predicates.iter().find_map(|(pred, _)| {
+                let pred = if let Some(substs) = substs {
+                    pred.subst(tcx, substs).kind().skip_binder()
+                } else {
+                    pred.kind().skip_binder()
+                };
+                if let ty::PredicateKind::Trait(pred) = pred && pred.self_ty() == ty {
+                            if Some(pred.def_id()) == tcx.lang_items().fn_trait() {
+                                return Some(hir::Mutability::Not);
+                            } else if Some(pred.def_id()) == tcx.lang_items().fn_mut_trait() {
+                                return Some(hir::Mutability::Mut)
+                            }
+                        }
+                None
```

Comment:
looks like formatting is broken because of let chains here

```suggestion
                if let ty::PredicateKind::Trait(pred) = pred && pred.self_ty() == ty {
                    if Some(pred.def_id()) == tcx.lang_items().fn_trait() {
                        Some(hir::Mutability::Not)
                    } else if Some(pred.def_id()) == tcx.lang_items().fn_mut_trait() {
                        Some(hir::Mutability::Mut)
                    }
                } else {
                    None
                }
```